File: paas-ce/paasagent/server/http.go
// Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
// Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
// Licensed under the MIT License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://opensource.org/licenses/MIT
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package server

import (
	"fmt"
	"log"
	"strconv"

	"github.com/labstack/echo"
	"github.com/labstack/echo/engine/standard"
	"github.com/spf13/viper"

	"paasagent/core"
	"paasagent/job"
)

// DeployData used for app deployment
type DeployData struct {
	AppCode      string                 `json:"app_code"`
	Mode         string                 `json:"mode"`
	Envs         map[string]interface{} `json:"envs"`
	EventID      string                 `json:"event_id"`
	DeployToken  string                 `json:"deploy_token"`
	DeployVars   map[string]interface{} `json:"deploy_vars"`
	SaaSSettings map[string]interface{} `json:"saas_settings"`
}

func validateAgentAuth(c echo.Context) (bool, string) {
	XID := c.Request().Header().Get("X-ID")
	XTOKEN := c.Request().Header().Get("X-TOKEN")

	sid := viper.GetString("auth.sid")
	token := viper.GetString("auth.token")

	if sid == XID && token == XTOKEN {
		return true, ""
	}
	return false, "sid or token is not valid, please check"
}

func controllerStatus(controllerServerURL string) error {
	url := controllerServerURL + "/v1/healthz"
	_, _, errs := core.DoGet(url)
	return errs
}

func appJobHandler(c echo.Context, handle string) error {
	valid, msg := validateAgentAuth(c)
	if !valid {
		log.Println(msg)
		return ForbiddenResponse(c, msg)
	}

	jobData := new(DeployData)
	if err := c.Bind(jobData); err != nil {
		return JsonResponse(c, map[string]interface{}{"error": 1, "msg": "envs params loads error"})
	}

	go func(handle string) {
		log.Println("RunJob start ... ...")
		if handle == "OFF" {
			appJob := job.AppJob{
				AppCode:      jobData.AppCode,
				Mode:         jobData.Mode,
				Envs:         map[string]interface{}{"BK_VERSION": "2.0"},
				EventID:      jobData.EventID,
				DeployToken:  jobData.DeployToken,
				DeployVars:   jobData.DeployVars,
				SaaSSettings: jobData.SaaSSettings,
				Handle:       handle,
			}
			appJob.OfflineJob()
		} else {
			appJob := job.AppJob{
				AppCode:      jobData.AppCode,
				Mode:         jobData.Mode,
				Envs:         jobData.Envs,
				EventID:      jobData.EventID,
				DeployToken:  jobData.DeployToken,
				DeployVars:   jobData.DeployVars,
				SaaSSettings: jobData.SaaSSettings,
				Handle:       handle,
			}
			appJob.OnlineJob()
		}
	}(handle)

	return JsonResponse(c, map[string]int{"error": 0})
}

func appOnlineHandler(c echo.Context) error {
	return appJobHandler(c, "ON")
}

func appOfflineHandler(c echo.Context) error {
	return appJobHandler(c, "OFF")
}

func healthzHandler(c echo.Context) error {
	valid, msg := validateAgentAuth(c)
	if !valid {
		log.Println(msg)
		return ForbiddenResponse(c, msg)
	}
	macAddrs, _ := core.GetMacAddrs()
	if len(macAddrs) > 0 {
		return JsonResponse(c, map[string]interface{}{"error": 0, "mac": macAddrs[0]})
	}
	return JsonResponse(c, map[string]interface{}{"error": 0, "mac": ""})
}

func healthzCheckHandler(c echo.Context) error {
	return StringResponse(c, "paasagent is alive")
}

func Run() error {
	controllerServerURL := viper.GetString("settings.CONTROLLER_SERVER_URL")
	if err := controllerStatus(controllerServerURL); err != nil {
		return fmt.Errorf("paasagent start failed, please make sure %s has been enabled, "+
			"and normal access to the service from this machine can be achieved.\n%s",
			controllerServerURL, err)
	}

	e := echo.New()
	e.POST("/v1/app/online", appOnlineHandler)
	e.POST("/v1/app/offline", appOfflineHandler)
	e.GET("/v1/app/healthz", healthzHandler)
	// check if paasagent is alive
	e.GET("/healthz", healthzCheckHandler)

	port := viper.GetInt("port")
	ip := viper.GetString("ip")
	log.Println("Start agent Now listen on ", ip, port)
	return e.Run(standard.New(ip + ":" + strconv.Itoa(port)))
}

File: paas2/appengine/api/http.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS
Community Edition) available.
Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

请求app engine的底层方法

Rules:
1. POST/DELETE/PUT: json in - json out, 如果resp.json报错, 则是engine接口问题
2. GET带参数 HEAD不带参数
3. 以统一的header头发送请求
"""

import requests

# from common.log import logging
import logging


def _gen_header(sid, token):
    headers = {"Content-Type": "application/json", "X-ID": str(sid), "X-TOKEN": str(token)}
    return headers


def _http_request(method, url, headers=None, data=None, timeout=None):
    try:
        if method == "GET":
            resp = requests.get(url=url, headers=headers, params=data, timeout=timeout)
        elif method == "HEAD":
            resp = requests.head(url=url, headers=headers)
        elif method == "POST":
            resp = requests.post(url=url, headers=headers, json=data)
        elif method == "DELETE":
            resp = requests.delete(url=url, headers=headers, json=data)
        elif method == "PUT":
            resp = requests.put(url=url, headers=headers, json=data)
        else:
            return False, None
    except requests.exceptions.RequestException:
        logging.exception("engine http request error! type: %s, url: %s, data: %s" % (method, url, str(data)))
        return False, None
    except requests.exceptions.Timeout:
        logging.exception("engine http request time out! type: %s, url: %s, data: %s" % (method, url, str(data)))
        return False, None
    else:
        if resp.status_code != 200:
            if resp.status_code == 403:
                return True, resp.json()
            logging.error(
                (
                    "engine http request error! type: %s, url: %s, data: %s, "
                    "response_status_code: %s, response_content: %s"
                )
                % (method, url, str(data), resp.status_code, resp.content)
            )
            return False, None

        return True, resp.json()


def http_get(url, sid, token, params, timeout=None):
    headers = _gen_header(sid, token)
    return _http_request(method="GET", url=url, headers=headers, data=params, timeout=timeout)


def http_post(url, sid, token, params):
    headers = _gen_header(sid, token)
    return _http_request(method="POST", url=url, headers=headers, data=params)
